dds/idl/field_info.cpp

```diff
@@ -0,0 +1,202 @@
+      switch (p->pt()) {
+      case AST_PredefinedType::PT_long: elem_sz_ = 4; elem_ = "ACE_CDR::Long"; return;
+      case AST_PredefinedType::PT_ulong: elem_sz_ = 4; elem_ = "ACE_CDR::ULong"; return;
+      case AST_PredefinedType::PT_longlong: elem_sz_ = 8; elem_ = "ACE_CDR::LongLong"; return;
+      case AST_PredefinedType::PT_ulonglong: elem_sz_ = 8; elem_ = "ACE_CDR::ULongLong"; return;
+      case AST_PredefinedType::PT_short: elem_sz_ = 2; elem_ = "ACE_CDR::Short"; return;
+      case AST_PredefinedType::PT_ushort: elem_sz_ = 2; elem_ = "ACE_CDR::UShort"; return;
+      case AST_PredefinedType::PT_float: elem_sz_ = 4; elem_ = "ACE_CDR::Float"; return;
+      case AST_PredefinedType::PT_double: elem_sz_ = 8; elem_ = "ACE_CDR::Double"; return;
+      case AST_PredefinedType::PT_longdouble: elem_sz_ = 16; elem_ = "ACE_CDR::LongDouble"; return;
+      case AST_PredefinedType::PT_char: elem_sz_ = 1; elem_ = "ACE_CDR::Char"; return;
+      case AST_PredefinedType::PT_wchar: elem_sz_ = 1; elem_ = "ACE_CDR::WChar"; return; // encoding of wchar length is 1 byte
+      case AST_PredefinedType::PT_boolean: elem_sz_ = 1; elem_ = "ACE_CDR::Boolean"; return;
+      case AST_PredefinedType::PT_octet: elem_sz_ = 1; elem_ = "ACE_CDR::Octet"; return;
+      default: break;
+      }
+    }
+    elem_ = scoped(as_act_->name());
+  }
+}
+
+std::string FieldInfo::string_type(Classification c)
+{
+  return be_global->language_mapping() == BE_GlobalData::LANGMAP_CXX11 ?
+    ((c & CL_WIDE) ? "std::wstring" : "std::string") :
+    (c & CL_WIDE) ? "TAO::WString_Manager" : "TAO::String_Manager";
+}
+
+std::string FieldInfo::to_cxx_type(AST_Type* type, std::size_t& size)
+{
+  const Classification cls = classify(type);
+  if (cls & CL_ENUM) { size = 4; return "ACE_CDR::ULong"; }
+  if (cls & CL_STRING) { size = 4; return string_type(cls); } // encoding of str length is 4 bytes
+  if (cls & CL_PRIMITIVE) {
+    type = resolveActualType(type);
+    AST_PredefinedType* p = AST_PredefinedType::narrow_from_decl(type);
+    switch (p->pt()) {
+    case AST_PredefinedType::PT_long: size = 4; return "ACE_CDR::Long";
+    case AST_PredefinedType::PT_ulong: size = 4; return "ACE_CDR::ULong";
+    case AST_PredefinedType::PT_longlong: size = 8; return "ACE_CDR::LongLong";
+    case AST_PredefinedType::PT_ulonglong: size = 8; return "ACE_CDR::ULongLong";
+    case AST_PredefinedType::PT_short: size = 2; return "ACE_CDR::Short";
+    case AST_PredefinedType::PT_ushort: size = 2; return "ACE_CDR::UShort";
+    case AST_PredefinedType::PT_float: size = 4; return "ACE_CDR::Float";
+    case AST_PredefinedType::PT_double: size = 8; return "ACE_CDR::Double";
+    case AST_PredefinedType::PT_longdouble: size = 16; return "ACE_CDR::LongDouble";
+    case AST_PredefinedType::PT_char: size = 1; return "ACE_CDR::Char";
+    case AST_PredefinedType::PT_wchar: size = 1; return "ACE_CDR::WChar"; // encoding of wchar length is 1 byte
```

Comment:
Yeah, at first I thought it was a typo, but the comment indicated otherwise. 
      case AST_PredefinedType::PT_wchar:
        size = 1; // encoding of wchar length is 1 byte
        return "ACE_CDR::WChar";

Then later I forgot to ask this question. @mitza-oci could you please verify this?

Reply:
Adam confirmed it's correct.

Reply:
See Serializer.inl `operator<<(Serializer&, ACE_OutputCDR::from_wchar)`
Note that the comment is talking about how the "length of the wchar" is encoded (not the wchar itself)

Reply:
I know about the length byte because I removed it in #1668. I guess a better question to ask is how is this size of one being used? Right now glancing at the code it looks like we don't use it, so unless I missing something, so why are we saying the size is 1 byte?

Also I'm this is something I missed in #1668, so more work probably needs to be done on the two byte wchar in XTypes, at least in `metaclass_generator::gen_typedef`.

Reply:
It is used in gen_field_getValueFromSerialized (metaclass generator).  Scalar types that are not strings but are wide (that is, wchar) have an additional block of code generated for "skip over."

Reply:
It doesn't use size there, I want to know why does `to_cxx_type` set the size of wchar at all.

Going off of `metaclass_generator.cpp` on master, this is how call the functions that can call `to_cxx_type` behave when the type is a scalar wchar:

- `gen_field_getValueFromSerialized` calls `to_cxx_type` doesn't use size unless it's a scalar that's not a string or wide.
- `gen_typedef` doesn't call  `to_cxx_type`.
- The `gen_union` callback for `generateSwitchForUnion`, `func` also doesn't call `to_cxx_type`.

So the answer is that it doesn't use the `size = 1` at all. It might have been used (or envisioned to be used) at some point, but it's confusing because you have to get the size of the wchar regardless. It looks like it does a similar thing for string, but it doesn't look like that's used either. Both of these should act like the default case, which doesn't set the size because the size isn't actually known.

XTypes will fix this by making wchar a fixed primitive again, so it will need the size, but I say the `size = 4` should be removed from the string case. Maybe the size information should be broken out into a `scalar_primitive_serialized_size` function.

Reply:
Yes, I see the point with 'size' here being potentially confusing.  Some of that is beyond the scope of this PR.

For this PR what I'm going to suggest is to move `to_cxx_type` back to `metaclass_generator.cpp` since that is the only place it is used.  Each `*_generator.cpp` is its own "pass" over the AST that has its own local concerns.  These don't need to be moved to other files unless they are needed by multiple generators.

Reply:
to_cxx_type is used in FieldInfo to set elem_ and elem_sz_. The duplicated method set_element() is removed, since to_cxx_type is used for both Field and element. That's the reason to_cxx_type is taken out of its original compilation unit scope. If we factor out all the common code and use FieldInfo wherever appropriate, to_cxx_type may not need to be static. Also we would have all the c++11 (or future) checks wrapped in one class.

Reply:
I went through it yesterday and tracked where the outputs of to_cxx_type were used, and it was only for metaclass_generator.  That resulted in my request that it be moved back there.